pkg/cloudproviders/aws/metadata.go

```diff
@@ -114,3 +123,56 @@ func plaintextIdentityDoc(ctx context.Context, mdClient *ec2metadata.EC2Metadata
 
 	return doc, nil
 }
+
+// getClusterMetadata attempts to get the EKS cluster name on a best effort basis.
+// First, it tries to get the cluster name from the EC2 instance tags. Access to
+// the tags must be explicitly enabled for the EC2 instance beforehand.
+// Second, it tries the node labels. The label is only set when the EKS cluster
+// was created via eksctl.
+func getClusterMetadata(ctx context.Context,
+	client *ec2metadata.EC2Metadata, doc *ec2metadata.EC2InstanceIdentityDocument,
+) *storage.ClusterMetadata {
+	clusterName, err := getClusterNameFromInstanceTags(ctx, client)
+	if err == nil {
+		return clusterMetadataFromName(clusterName, doc)
+	}
+	log.Errorf("Failed to get EKS cluster metadata from instance tags: %v", err)
+
+	config, err := k8sutil.GetK8sInClusterConfig()
+	if err != nil {
+		log.Errorf("Obtaining in-cluster Kubernetes config: %v", err)
+		return nil
+	}
+	k8sClient := k8sutil.MustCreateK8sClient(config)
+	clusterName, err = getClusterNameFromNodeLabels(ctx, k8sClient)
+	if err == nil {
+		return clusterMetadataFromName(clusterName, doc)
+	}
+	log.Errorf("Failed to get EKS cluster metadata from node labels: %v", err)
+	return nil
```

Comment:
For readability, let's invert that one and log `if err != nil` instead and by default return `clusterMetadataFromName()`.